marlowe-runtime/sync/Language/Marlowe/Runtime/Sync/Database/PostgreSQL/GetContractState.hs

```diff
@@ -0,0 +1,292 @@
+      LEFT JOIN marlowe.txOutAsset USING (txId, txIx)
+      WHERE createTxOut.txId = $1 :: bytea
+        AND createTxOut.txIx = $2 :: smallint
+      GROUP BY createTxOut.txId, createTxOut.txIx
+    |]
+  (latestBlock, latestOutput) <- lift $ maybe (initialBlock, Just initialOutput) decodeLatestResults <$> T.statement params [maybeStatement|
+    SELECT
+      applyTx.txId :: bytea,
+      (ARRAY_AGG(applyTx.outputTxIx))[1] :: smallint?,
+      (ARRAY_AGG(applyTx.slotNo))[1] :: bigint,
+      (ARRAY_AGG(applyTx.blockId))[1] :: bytea,
+      (ARRAY_AGG(applyTx.blockNo))[1] :: bigint,
+      (ARRAY_AGG(contractTxOut.rolesCurrency))[1] :: bytea?,
+      (ARRAY_AGG(contractTxOut.state))[1] :: bytea?,
+      (ARRAY_AGG(contractTxOut.contract))[1] :: bytea?,
+      (ARRAY_AGG(txOut.address))[1] :: bytea?,
+      (ARRAY_AGG(txOut.lovelace))[1] :: bigint?,
+      ARRAY_REMOVE(ARRAY_AGG(txOutAsset.policyId), NULL) :: bytea[],
+      ARRAY_REMOVE(ARRAY_AGG(txOutAsset.name), NULL) :: bytea[],
+      ARRAY_REMOVE(ARRAY_AGG(txOutAsset.quantity), NULL) :: bigint[]
+    FROM marlowe.applyTx
+    LEFT JOIN marlowe.contractTxOut
+      ON contractTxOut.txId = applyTx.txId
+      AND contractTxOut.txIx = applyTx.outputTxIx
+    LEFT JOIN marlowe.txOut
+      ON txOut.txId = applyTx.txId
+      AND txOut.txIx = applyTx.outputTxIx
+    LEFT JOIN marlowe.txOutAsset
+      ON txOutAsset.txId = applyTx.txId
+      AND txOutAsset.txIx = applyTx.outputTxIx
+    LEFT JOIN marlowe.applyTx AS consumer
+      ON consumer.inputTxId = applyTx.txId
+      AND consumer.inputTxIx = applyTx.outputTxIx
+    WHERE applyTx.createTxId = $1 :: bytea
+      AND applyTx.createTxIx = $2 :: smallint
+      AND consumer.txId IS NULL
+    GROUP BY applyTx.txId
+  |]
+  availablePayouts <- lift
+    $ Map.fromDistinctAscList . V.toList . fmap decodePayout
+    <$> T.statement params [vectorStatement|
+      SELECT
+        payoutTxOut.txId :: bytea,
+        payoutTxOut.txIx :: smallint,
+        (ARRAY_AGG(payoutTxOut.rolesCurrency))[1] :: bytea,
+        (ARRAY_AGG(payoutTxOut.role))[1] :: bytea,
+        (ARRAY_AGG(txOut.address))[1] :: bytea,
+        (ARRAY_AGG(txOut.lovelace))[1] :: bigint,
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.policyId), NULL) :: bytea[],
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.name), NULL) :: bytea[],
+        ARRAY_REMOVE(ARRAY_AGG(txOutAsset.quantity), NULL) :: bigint[]
+      FROM marlowe.payoutTxOut
+      JOIN marlowe.txOut USING (txId, txIx)
+      LEFT JOIN marlowe.txOutAsset USING (txId, txIx)
+      JOIN marlowe.applyTx USING (txId)
+      LEFT JOIN marlowe.withdrawalTxIn
+        ON withdrawalTxIn.payoutTxId = payoutTxOut.txId
+        AND withdrawalTxIn.payoutTxIx = payoutTxOut.txIx
+      WHERE applyTx.createTxId = $1 :: bytea
+        AND applyTx.createTxIx = $2 :: smallint
+        AND withdrawalTxIn.txId IS NULL
+      GROUP BY payoutTxOut.txId, payoutTxOut.txIx
+      ORDER BY payoutTxOut.txId, payoutTxOut.txIx
+    |]
```

Comment:
We actually have a choice here between two behaviors, in situations where several contracts use the same policy ID for role tokens.
1. This query only returns the available payouts resulting from this particular contract, which is what happens here.
2. This query returns all payouts from the contract's policy ID.

Eventually, we might want a separate query that provides the second result, since that would be convenient for wallet-centric use cases.